src/rm_cms.c

```diff
@@ -0,0 +1,328 @@
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+
+    if(parseCreateArgs(ctx, argv, argc, &width, &depth) != REDISMODULE_OK)
+        return REDISMODULE_ERR;
+
+    if(RedisModule_KeyType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx,"CMS: key already exists");
+    }
+
+    CreateCMSKey(ctx, keyName, width, depth, &cms, &key);
+
+    RedisModule_CloseKey(key);
+    RedisModule_Log(ctx, "info", "created new count-min sketch");
+    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    RedisModule_ReplicateVerbatim(ctx);
+    return REDISMODULE_OK;
+}
+
+static int parseIncrByArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
+                CMSPair **pairs, int qty) {
+    for(int i = 0; i < qty; ++i) {
+        (*pairs)[i].key = RedisModule_StringPtrLen(argv[2 + i * 2], NULL);
+        RedisModule_StringToLongLong(argv[2 + i * 2 + 1], &((*pairs)[i].value));
+    }
+    return REDISMODULE_OK;
+}
+
+int CMSketch_incrBy(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+
+    if(argc < 4 || (argc % 2) == 1) {
+        return RedisModule_WrongArity(ctx);
+    }
+
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ);
+    CMSketch *cms = NULL;
+
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        CreateCMSKey(ctx, keyName, DEFAULT_WIDTH, DEFAULT_DEPTH, &cms, &key);
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+    } else {
+        cms = RedisModule_ModuleTypeGetValue(key);
+    }
+ 
+    int pairCount = (argc - 2) / 2;
+    CMSPair *pairArray= CMS_CALLOC(pairCount, sizeof(CMSPair));
+    parseIncrByArgs(ctx, argv, argc, &pairArray, pairCount);
+    for(int i = 0; i < pairCount; ++i) {
+        CMS_IncrBy(cms, pairArray[i].key, pairArray[i].value);
+    }    
+
+    CMS_FREE(pairArray);
+    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    RedisModule_CloseKey(key);
+    return REDISMODULE_OK;
+}
+
+int CMSketch_query(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+
+    if(argc < 3) {
+        return RedisModule_WrongArity(ctx);
+    }
+
+    CMSketch *cms = NULL;
+    if(GetCMSKey(ctx, argv[1], &cms, REDISMODULE_READ) != REDISMODULE_OK) {
+        return REDISMODULE_OK;
+    }
+
+    int itemCount = argc - 2;
+    size_t length = 0;
+    RedisModule_ReplyWithArray(ctx, itemCount);
+    for(int i = 0; i < itemCount; ++i) {
+        const char *str = RedisModule_StringPtrLen(argv[2 + i], &length);
+        RedisModule_ReplyWithLongLong(ctx, CMS_Query(cms, str));
+    }
+
+    return REDISMODULE_OK;
+}
+
+static int parseMergeArgs(RedisModuleCtx *ctx, RedisModuleString **argv,
```

Comment:
suggest that you use a structur e.g. `typedef struct { ... } mergeParams`

Reply:
Done

Reply:
Done